File: src/api/achievements.js
import { invoke } from "@tauri-apps/api/core";

// Echtes Achievement-Backend (23.08.2026, siehe AchievementController.php +
// achievements.rs/achievements_commands.rs auf der Tauri-Seite) - ersetzt
// den bisherigen reinen Client-Mock. Gleiche Feldnamen/async-Signatur wie
// vorher, damit AchievementsScreen.jsx unverändert bleiben konnte.
//
// `condition`/interne Tracking-Logik bleiben serverseitig - der Launcher
// bekommt nur das fertige JSON (id/category/step/title/description/icon/
// unlocked/unlockedAt/contextSentence/percentUnlocked/progressPercent/
// justUnlocked), siehe Backend-Kommentar in AchievementController.php.

export async function getStats() {
  const stats = await invoke("get_achievement_stats");
  trackForNewUnlocks(await getAchievementsQuiet());
  return stats;
}

export async function getAchievements() {
  const achievements = await invoke("get_achievements");
  trackForNewUnlocks(achievements);
  return achievements;
}

/** Wie getAchievements(), aber ohne den Unlock-Tracking-Seiteneffekt
 * auszulösen - genutzt intern von getStats(), damit dort nicht doppelt
 * getrackt wird, falls beide kurz hintereinander aufgerufen werden. */
async function getAchievementsQuiet() {
  return invoke("get_achievements");
}

// --- Polling-Ersatz für den früheren rein clientseitigen Unlock-Timer ---
//
// Der Mock hatte einen simulierten Freischalt-Moment während der laufenden
// Session (scheduleSimulatedUnlock). Das echte Backend hat KEIN Push-/
// Event-System für neue Freischaltungen (die schreibt der
// achievements:check-Scheduler alle 15 Minuten serverseitig weg) - deshalb
// wird hier stattdessen periodisch neu abgerufen und mit dem zuletzt
// bekannten Freischalt-Zustand verglichen. Ein Wechsel unlocked:false ->
// unlocked:true (oder ein frisches justUnlocked:true) zwischen zwei Abrufen
// gilt als "neu freigeschaltet" und löst denselben Frisch-geschmiedet-Effekt
// aus wie vorher der Mock-Timer.
const POLL_INTERVAL_MS = 60_000;

let lastKnownUnlocked = null; // Map<id, bool> | null (null = noch kein Abruf)
const listeners = new Set();
let pollTimerStarted = false;

function trackForNewUnlocks(achievements) {
  const previous = lastKnownUnlocked;
  const current = new Map(achievements.map((a) => [a.id, a.unlocked]));

  if (previous) {
    for (const a of achievements) {
      const wasUnlocked = previous.get(a.id) ?? false;
      if (a.unlocked && !wasUnlocked) {
        listeners.forEach((fn) => fn({ ...a, justUnlocked: true }));
      } else if (a.unlocked && a.justUnlocked) {
        // Bereits freigeschaltet, aber serverseitig noch nicht bestätigt
        // (seen_at NULL) - z. B. weil der Nutzer die Schmiede seit der
        // Freischaltung nicht mehr geöffnet hat. Trotzdem als "neu" melden.
        listeners.forEach((fn) => fn({ ...a }));
      }
    }
  }

  lastKnownUnlocked = current;
}

function schedulePolling() {
  if (pollTimerStarted) return;
  pollTimerStarted = true;
  window.setInterval(() => {
    if (listeners.size === 0) return;
    getAchievements().catch(() => {
      // Netzwerkfehler beim Hintergrund-Poll - kein kritischer Pfad, nächster
      // Poll versucht es erneut.
    });
  }, POLL_INTERVAL_MS);
}

/** Feuert, sobald ein Achievement neu freigeschaltet erkannt wird (entweder
 * durch einen Wechsel seit dem letzten Abruf, oder weil der letzte Abruf
 * bereits ein noch unbestätigtes justUnlocked:true zurückgab). */
export function subscribeNewUnlock(callback) {
  listeners.add(callback);
  schedulePolling();
  return () => listeners.delete(callback);
}

/** Markiert ein frisch freigeschaltetes Achievement als "gesehen" (setzt
 * seen_at serverseitig), damit der Frisch-geschmiedet-Effekt nur einmal
 * abläuft. Fire-and-forget - kein kritischer Pfad für die Anzeige. */
export function acknowledgeJustUnlocked(id) {
  invoke("acknowledge_achievement", { achievementId: id }).catch(() => {
    // Bestätigung fehlgeschlagen (z. B. Netzwerk) - der Effekt läuft dann
    // beim nächsten Öffnen der Schmiede erneut ab, kein Datenverlust.
  });
}

File: src-tauri/src/achievements.rs
//! Echtes Achievement-Backend (23.08.2026) über die Sanctum-authentifizierten
//! app-api/achievements-Endpunkte (siehe AchievementController.php auf dem
//! Server) - ersetzt den bisherigen reinen Client-Mock in `achievements.js`.
//! Nutzt denselben Sanctum-Token-Mechanismus wie `social.rs`
//! (`social::ensure_sanctum_token`), daher kein eigenes Token-Handling hier.

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

use crate::config;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AchievementEntry {
    pub id: String,
    pub category: String,
    pub step: i32,
    pub title: String,
    pub description: String,
    pub icon: String,
    pub unlocked: bool,
    #[serde(rename = "unlockedAt")]
    pub unlocked_at: Option<String>,
    #[serde(rename = "contextSentence")]
    pub context_sentence: Option<String>,
    #[serde(rename = "percentUnlocked")]
    pub percent_unlocked: i32,
    #[serde(rename = "progressPercent")]
    pub progress_percent: i32,
    #[serde(rename = "justUnlocked")]
    pub just_unlocked: bool,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AchievementStats {
    #[serde(rename = "playTimeSeconds")]
    pub play_time_seconds: i64,
}

/// Liefert den Erfolge-Katalog inkl. Freischalt-/Fortschrittsstatus des
/// eingeloggten Spielers - identisches JSON-Shape wie der bisherige Mock
/// (`getAchievements()` in `achievements.js`).
pub async fn fetch_achievements(client: &reqwest::Client, sanctum_token: &str) -> Result<Vec<AchievementEntry>> {
    let resp = client
        .get(config::ERZMARK_ACHIEVEMENTS_URL)
        .bearer_auth(sanctum_token)
        .send()
        .await
        .context("Erfolge nicht erreichbar (Netzwerk?)")?;

    if !resp.status().is_success() {
        anyhow::bail!("Erfolge-Abruf fehlgeschlagen ({})", resp.status());
    }

    resp.json::<Vec<AchievementEntry>>()
        .await
        .context("Ungültige Antwort beim Abruf der Erfolge")
}

/// Aktuelle Spielzeit des eingeloggten Spielers - identisches JSON-Shape wie
/// der bisherige Mock (`getStats()` in `achievements.js`).
pub async fn fetch_achievement_stats(client: &reqwest::Client, sanctum_token: &str) -> Result<AchievementStats> {
    let resp = client
        .get(config::ERZMARK_ACHIEVEMENTS_STATS_URL)
        .bearer_auth(sanctum_token)
        .send()
        .await
        .context("Erfolge-Statistik nicht erreichbar (Netzwerk?)")?;

    if !resp.status().is_success() {
        anyhow::bail!("Erfolge-Statistik-Abruf fehlgeschlagen ({})", resp.status());
    }

    resp.json::<AchievementStats>()
        .await
        .context("Ungültige Antwort beim Abruf der Erfolge-Statistik")
}

/// Markiert einen frisch freigeschalteten Erfolg als gesehen (setzt
/// `seen_at` serverseitig), damit der Frisch-geschmiedet-Effekt im Frontend
/// nur einmal abläuft - Gegenstück zum bisherigen `acknowledgeJustUnlocked(id)`.
pub async fn acknowledge_achievement(client: &reqwest::Client, sanctum_token: &str, achievement_id: &str) -> Result<()> {
    let url = format!("{}/{}/ack", config::ERZMARK_ACHIEVEMENTS_ACK_URL_BASE, achievement_id);

    let resp = client
        .post(url)
        .bearer_auth(sanctum_token)
        .send()
        .await
        .context("Erfolg bestätigen fehlgeschlagen (Netzwerk?)")?;

    if !resp.status().is_success() {
        anyhow::bail!(
            "Erfolg bestätigen fehlgeschlagen ({}): {}",
            resp.status(),
            resp.text().await.unwrap_or_default()
        );
    }

    Ok(())
}
